### PR TITLE
composite-checkout: Refactor Checkout steps as direct children

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

We need a way to run step validation when the "Continue" button is pressed (as opposed to on keypress) and to be able to run that validation asynchronously. #39087 was an attempt to do that, but because of issues with scope and state access, it kept running into problems. It also further complicates the API for steps.

This PR takes a different approach by refactoring the way steps are provided to the checkout form which should hopefully both solve the scope issues plaguing the previous attempt while at the same time making the step API easier to understand and use.

#### Background

Steps were conceived as "step objects" which were plain JS objects that followed a defined schema. Most of that schema was a collection of React elements to be displayed in different parts of the step, and in different situations. That part worked fine. However, in order to run validation on the step's data, the object also included a function called `isCompleteCallback`.

The original idea for that function was something simple that could be called on every render and declaratively return if the step was complete, but it quickly became challenging to in practice since the data in the step components was not easily accessible in the scope where the step was defined, and redefining the step could cause performance issues. Furthermore, the WPCOM contact form performs its validation with a server call which we cannot reasonably call on every render even if we had the ability to call `isCompleteCallback` asynchronously.

#39087 attempted to solve these issues by calling `isCompleteCallback` only when the "Continue" button was pressed, but the scope issues still remained.

#### A new step API

In this PR we remove the concept of step objects entirely, and instead rely on React components. The checkout form still begins with the `Checkout` component, but now that component has children. Any child is allowed but the `CheckoutStepBody` component can be used to render something that looks like a checkout step. Rather than having to write your own logic for these steps, you can instead use the `CheckoutSteps` component wrapping `CheckoutStep` components. 

Most of the properties of the old step objects now reside as props on `CheckoutStep`. In fact, existing step objects can remain and their properties passed as these props since the underlying API remains the same.

The notable difference with the new API is that `isCompleteCallback`, now a prop, will not be called on every render and will instead be called with the "Continue" button is pressed. If the function returns a Promise, then the "Continue" button will change to "Please wait…" until the Promise resolves. This is the same approach taken in #39087 with one major advantage: since the callback is now defined in JSX, it has direct access to the scope of its parent, which can provide all the form data that the callback needs in its closure.

#### Testing instructions

Even though it's only at the top level, this is a pretty major change, so we'll need to verify that the form still behaves as it did before. Most notably we need to make sure that the steps change correctly, that active/complete/incomplete steps all display properly, that editing a step works as expected, and that validation works as expected.

The biggest visible change of this PR is that the "Continue" button is no longer disabled when a step is incomplete. Pressing it when the step has invalid data should display an error in some way.

This PR doesn't yet use async validation in Calypso, but it makes it possible. Try running the demo (`npm run composite-checkout-demo`) to see it in action.

#### To do

- [x] Update demo to use new API
- [x] Update WPCheckout to use new API
- [x] Make sure WPContactForm displays validation errors
- [x] Refactor tests to use new API
- [x] Update README with new API
- [x] Remove unused components and hooks from old API

### CODE REVIEW

**Comment by sirbrillig** [comment]
I still need to clean up old code, fix the tests and the documentation, but this is ready to be manually tested.

**Comment by nbloomf** [comment]
Thanks for working on this!

I don't think it's related to this PR, but I have run into a weird state in the demo. Steps to replicate are:
- Choose the PayPal payment method
- Choose "US" as the country
- Click on "Pay with PayPal" button
- The demo page freezes at this point, but I can still click the "edit" button on the payment method step. However, doing so puts me in a state where I can change my payment method, but the "Continue" button is never activated.

Probably what we need to do here is disable the edit button on the payment method and contact details steps once the payment button is in a processing state. I'll also check this in calypso.

**Comment by nbloomf** [comment]
Testing in Calypso now. I see a weird error when trying to complete a payment with Stripe:

image

This looks similar to an error we've seen before where there was a missing `value` accessor.

**Comment by fditrapani** [comment]
This is coming along! I can confirm the same issue as Nathan with being able to edit steps after buying with Paypal. I double checked in production and it doesn't behave this way. 

Another minor thing I noticed, is when you pick the credit card payment method, press continue without entering anything which shows the errors, then pick another payment method and continue. If you go back to the payment methods and pick the credit cards, the errors are still showing.  Is it possible to clear the errors the first time around if a valid payment method is picked?

File: packages/composite-checkout/src/components/checkout-steps.js [hunk 1]
```diff
@@ -31,15 +41,53 @@ export function Checkout( { children, className } ) {
...
+					setTotalSteps,
+				} }
+			>
+				<MainContentUI
+					className={ joinClasses( [ className, 'checkout__content' ] ) }
+					isLastStepActive={ isThereAnotherNumberedStep }
+				>
+					{ children }
+
+					<SubmittButtonWrapperUI isLastStepActive={ ! isThereAnotherNumberedStep }>
```
Comment (nbloomf) [inline packages/composite-checkout/src/components/checkout-steps.js]: Is there a reason why Submit is spelled with two 't's here?

**Comment by fditrapani** [comment]
Here's a weird error I ran into when submitting a stripe card while sandboxed:

image

image


I thought it was because I used a Canada as a country but I still see the error with US.

File: packages/composite-checkout/src/components/checkout-steps.js [hunk 2]
```diff
@@ -107,21 +111,23 @@ export function CheckoutSteps( { children } ) {
...
 	setStepCompleteStatus,
 } ) {
+	const localize = useLocalize();
+	const { totalSteps } = useContext( CheckoutStepDataContext );
 	const { setFormPending, setFormReady } = useFormStatus();
 	const goToThisStep = () => setActiveStepNumber( stepNumber );
 	const goToNextStep = () => {
-		const completeResult = isComplete();
+		const completeResult = isCompleteCallback();
 		if ( completeResult.then ) {
```
Comment (nbloomf) [inline packages/composite-checkout/src/components/checkout-steps.js]: Is this a reliable way to detect whether something is a Promise? (I really don't know)

File: packages/composite-checkout/demo/index.js [hunk 3]
```diff
@@ -329,6 +331,17 @@ function MyCheckout() {
 			paymentMethods={ [ applePayMethod, stripeMethod, paypalMethod ].filter( Boolean ) }
 		>
 			<Checkout>
+				<CheckoutStepBody
+					activeStepContent={ orderSummaryStep.activeStepContent }
+					completeStepContent={ orderSummaryStep.completeStepContent }
+					titleContent={ orderSummaryStep.titleContent }
+					errorMessage={ 'There was an error with this step.' }
```
Comment (nbloomf) [inline packages/composite-checkout/demo/index.js]: Does this string need to be localized?

File: packages/composite-checkout/src/components/checkout-steps.js [hunk 4]
```diff
@@ -217,7 +217,7 @@ export function CheckoutStepBody( {
 					title={ titleContent }
 					isActive={ isStepActive }
 					isComplete={ isStepComplete }
-					onEdit={ isStepComplete && goToThisStep ? goToThisStep : null }
+					onEdit={ isStepComplete && goToThisStep && ! isStepActive ? goToThisStep : null }
```
Comment (nbloomf) [inline packages/composite-checkout/src/components/checkout-steps.js]: I think this is where we'd disable editing when the form is in submitting state. If you think it makes sense we can include that change here.

File: packages/composite-checkout/src/lib/payment-methods/stripe-credit-card-fields.js [hunk 5]
```diff
@@ -199,7 +200,7 @@ export function createStripeMethod( {
 						transactionStatus: null,
 					};
 				case 'CARDHOLDER_NAME_SET':
-					return { ...state, cardholderName: action.payload };
+					return { ...state, cardholderName: { value: action.payload, isTouched: true } };
```
Comment (nbloomf) [inline packages/composite-checkout/src/lib/payment-methods/stripe-credit-card-fields.js]: This change might be related to the Stripe error reported above.

File: packages/composite-checkout/demo/index.js (hunk 3 above)
Comment (nbloomf) [reply packages/composite-checkout/demo/index.js]: Ah- I see this is in the demo. So no :)

File: packages/composite-checkout/src/components/checkout-steps.js (hunk 1 above)
Comment (sirbrillig) [reply packages/composite-checkout/src/components/checkout-steps.js]: Haha, my own haste. I will fix that; thanks.

File: packages/composite-checkout/src/components/checkout-steps.js (hunk 2 above)
Comment (sirbrillig) [reply packages/composite-checkout/src/components/checkout-steps.js]: I actually did some research, and this is the _only_ proper way to detect if something is a promise! More specifically, we are checking if something is "thennable".

**Comment by sirbrillig** [comment]
> The demo page freezes at this point,

I can fix this once #39654 is merged, but it would introduce a lot of conflicts if I tried to do it now.

> what we need to do here is disable the edit button on the payment method and contact details steps once the payment button is in a processing state

good point! Done.

> This looks similar to an error we've seen before where there was a missing value accessor.

You're right. I missed pulling the value out of the object when it was submitted. Fixed.

**Comment by sirbrillig** [comment]
I think I finally fixed all the tests 😥

**Comment by sirbrillig** [comment]
Rebased on master and removed the old code this replaces. The only thing left I believe is the README. Please test again when you have a chance!

**Comment by sirbrillig** [comment]
Rebased after #39654

**Comment by sirbrillig** [comment]
> The demo page freezes at this point,

That should be fixed now as well. This is ready for review once again.

**Comment by sirbrillig** [comment]
I had forgotten to restore the url hash code. It's back.

**Comment by nbloomf** [comment]
I went on a shopping spree, testing every combination of
- new card, new card + credits, stored card, stored card + credits, paypal, paypal + credits, full credits
with
- wpcom signup, wpcom upgrade

and did not see any problems. Editing completed steps works as expected, and I got validation errors when trying to submit empty contact fields.

**Pull request merged**